src/server-core/src/lib.rs

```diff
@@ -69,6 +70,52 @@ impl Connection {
     pub fn uuid_handle(&self) -> ConnectionUuidHandle {
         ConnectionUuidHandle(Arc::clone(&self.conn_uuid))
     }
+
+    /// Attempts to parse a proxy header from the tcp_stream.
+    /// If none is found or it is unable to be parsed None will
+    /// be returned. If a header is found it will be returned and its
+    /// bytes will be removed from the stream.
+    ///
+    /// It is possible an invalid header was sent, if that is the case
+    /// any downstream service will be responsible for returning errors
+    /// to the client.
+    pub async fn take_proxy_header(&mut self) -> Option<ProxyHeader> {
+        // 1024 bytes is a rather large header for tcp proxy header, unless
+        // if the header contains TLV fields or uses a unix socket address
+        // this could easily be hit. We'll use a 1024 byte max buf to allow
+        // limited support for this.
+        let mut buf = [0u8; 1024];
+        let len = match self.tcp_stream.peek(&mut buf).await {
+            Ok(n) if n > 0 => n,
+            _ => {
+                debug!("Failed to read from client socket or no data received");
+                return None;
+            }
+        };
+
+        // Attempt to parse the header, and log failures.
+        let (header, hlen) = match ProxyHeader::parse(&buf[..len], ParseConfig::default()) {
+            Ok((header, hlen)) => (header, hlen),
+            Err(proxy_header::Error::Invalid) => {
+                debug!(
+                    "Proxy header is invalid. This is likely due to no no header being provided",
+                );
+                return None;
+            }
+            Err(e) => {
+                error!("Proxy header parse error '{:?}', ignoring header.", e);
+                return None;
+            }
+        };
+        debug!("Proxied connection with header {:?}", header);
+        // Proxy header found, clear the bytes
+        let _ = self.read_exact(&mut vec![0; hlen]).await;
+        Some(ProxyHeader::into_owned(header))
```

Comment:
If you return the `ProxiedAddress`, you can just use the existing buffer to actually read the bytes:

```suggestion
        let address = header.proxied_address();
        let _ = self.read_exact(&mut buf[..hlen]).await;
        Some(address)
```

Reply:
Great idea, and very close!
```rust
      let address = header.proxied_address().map(|a| a.to_owned());
      let _ = self.read_exact(&mut buf[..hlen]).await;
      address
```